cpp/tests/listener.cpp

```diff
@@ -174,34 +191,163 @@ TEST_F(ListenerTest, RaiseOnError)
+  auto ep =
+    _worker->createEndpointFromHostname("127.0.0.1", listener->getPort(), _endpointErrorHandling);
 
-  bool isClosed = false;
-  ep->setCloseCallback([](void* isClosed) { *reinterpret_cast<bool*>(isClosed) = true; },
-                       reinterpret_cast<void*>(&isClosed));
+  struct CallbackData {
+    ucs_status_t status{UCS_INPROGRESS};
+    bool closed{false};
+  };
+
+  auto callbackData = std::make_shared<CallbackData>();
+  ep->setCloseCallback(
+    [](ucs_status_t status, ucxx::EndpointCloseCallbackUserData callbackData) {
+      auto cbData    = std::static_pointer_cast<CallbackData>(callbackData);
+      cbData->status = status;
+      cbData->closed = true;
+    },
+    callbackData);
 
   while (listenerContainer->endpoint == nullptr)
     _worker->progress();
 
-  ASSERT_FALSE(isClosed);
+  ASSERT_FALSE(callbackData->closed);
+  ASSERT_EQ(callbackData->status, UCS_INPROGRESS);
 
   listenerContainer->endpoint = nullptr;
 
-  loopWithTimeout(std::chrono::milliseconds(5000), [this, &isClosed]() {
+  loopWithTimeout(std::chrono::milliseconds(5000), [this, &callbackData]() {
     _worker->progress();
-    return isClosed;
+    return callbackData->closed;
   });
 
-  ASSERT_TRUE(isClosed);
+  ASSERT_TRUE(callbackData->closed);
+  EXPECT_NE(callbackData->status, UCS_INPROGRESS);
+}
+
+bool checkRequestWithTimeout(std::chrono::milliseconds timeout,
+                             std::shared_ptr<ucxx::Worker> worker,
+                             std::shared_ptr<ucxx::Request> closeRequest)
+{
+  auto startTime = std::chrono::system_clock::now();
+  auto endTime   = startTime + std::chrono::milliseconds(timeout);
+
+  while (std::chrono::system_clock::now() < endTime) {
+    worker->progress();
+    if (closeRequest->isCompleted()) return true;
+  }
+  return false;
 }
 
+TEST_P(ListenerTest, EndpointNonBlockingClose)
+{
+  auto listenerContainer = createListenerContainer();
+  auto listener          = createListener(listenerContainer);
+  _worker->progress();
+
+  auto ep =
+    _worker->createEndpointFromHostname("127.0.0.1", listener->getPort(), _endpointErrorHandling);
+
+  while (listenerContainer->endpoint == nullptr)
+    _worker->progress();
+
+  auto closeRequest = ep->close();
+
+  /**
+   * FIXME: For some reason the code below calls `_worker->progress()` from within
+   * `_worker->progress()`, which is invalid in UCX. The `checkRequestWithTimeout` below
+   * which is functionally equivalent has no such problem. The lambda seems to behave in
+   * unexpected way here. The issue also goes away if in `Endpoint::close()` the
+   * line `if (callbackFunction) callbackFunction(status, callbackData);` is commented
+   * out from the `combineCallbacksFunction` lambda, even when no callback is specified
+   * to `ep->close()` above.
+   */
+  // auto f = [this, &closeRequest]() {
+  //   _worker->progress();
+  //   return closeRequest->isCompleted();
+  // };
+  // loopWithTimeout(std::chrono::milliseconds(5000), f);
```

Comment:
Just uncommenting this and commenting out the code on line 280 didn't let me see the problem you describe...

Reply:
You should definitely see it if you have a UCX debug build. Just tried it now with UCX 1.15 debug and I'm able to reproduce it:

```
[dgx13:3842411:0:3842411]  ucp_worker.c:2888 Assertion `worker->inprogress++ == 0' failed

/src/ucx-1.15/build/src/ucp/../../../src/ucp/core/ucp_worker.c: [ ucp_worker_progress() ]
      ...
     2885     UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
     2886
     2887     /* check that ucp_worker_progress is not called from within ucp_worker_progress */
==>  2888     ucs_assert(worker->inprogress++ == 0);
     2889     count = uct_worker_progress(worker->uct);
     2890     ucs_async_check_miss(&worker->async);
     2891

==== backtrace (tid:3842411) ====
 0 0x000000000005fd79 ucp_worker_progress()  /src/ucx-1.15/build/src/ucp/../../../src/ucp/core/ucp_worker.c:2888
 1 0x000000000005bebe ucxx::Worker::progressOnce()  /src/ucxx/cpp/src/worker.cpp:254
 2 0x000000000005beec ucxx::Worker::progressPending()  /src/ucxx/cpp/src/worker.cpp:260
 3 0x000000000005ca07 ucxx::Worker::progress()  /src/ucxx/cpp/src/worker.cpp:268
 4 0x000000000005ca07 std::mutex::lock()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_mutex.h:100
 5 0x000000000005ca07 std::lock_guard<std::mutex>::lock_guard()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_mutex.h:229
 6 0x000000000005ca07 ucxx::Worker::progress()  /src/ucxx/cpp/src/worker.cpp:272
 7 0x000000000004e1c1 std::_Function_handler<bool (), (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()::{lambda()#2}>::_M_invoke()  /src/ucxx/cpp/tests/listener.cpp:318
 8 0x000000000004e1c1 std::__shared_ptr_access<ucxx::Request, (__gnu_cxx::_Lock_policy)2, false, false>::_M_get()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr_base.h:993
 9 0x000000000004e1c1 std::__shared_ptr_access<ucxx::Request, (__gnu_cxx::_Lock_policy)2, false, false>::operator->()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr_base.h:987
10 0x000000000004e1c1 operator()()  /src/ucxx/cpp/tests/listener.cpp:319
11 0x000000000004e1c1 __invoke_impl<bool, (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()::<lambda()>&>()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/invoke.h:61
12 0x000000000004e1c1 __invoke_r<bool, (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()::<lambda()>&>()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/invoke.h:114
13 0x000000000004e1c1 _M_invoke()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_function.h:290
14 0x000000000003ef43 std::function<void (ucs_status_t, std::shared_ptr<void>)>::operator()()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_function.h:590
15 0x000000000003ef43 std::__shared_ptr<void, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr_base.h:1154
16 0x000000000003ef43 std::shared_ptr<void>::~shared_ptr()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr.h:122
17 0x000000000003ef43 operator()()  /src/ucxx/cpp/src/endpoint.cpp:162
18 0x000000000003ef43 __invoke_impl<void, ucxx::Endpoint::close(bool, ucxx::EndpointCloseCallbackUserFunction, ucxx::EndpointCloseCallbackUserData)::<lambda(ucs_status_t, ucxx::EndpointCloseCallbackUserData)>&, ucs_status_t, std::shared_ptr<void> >()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/invoke.h:61
19 0x000000000003ef43 __invoke_r<void, ucxx::Endpoint::close(bool, ucxx::EndpointCloseCallbackUserFunction, ucxx::EndpointCloseCallbackUserData)::<lambda(ucs_status_t, ucxx::EndpointCloseCallbackUserData)>&, ucs_status_t, std::shared_ptr<void> >()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/invoke.h:111
20 0x000000000003ef43 _M_invoke()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_function.h:290
21 0x0000000000048c72 std::function<void (ucs_status_t, std::shared_ptr<void>)>::operator()()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_function.h:590
22 0x0000000000048c72 std::__shared_ptr<void, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr_base.h:1154
23 0x0000000000048c72 std::shared_ptr<void>::~shared_ptr()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr.h:122
24 0x0000000000048c72 ucxx::Request::setStatus()  /src/ucxx/cpp/src/request.cpp:230
25 0x0000000000049209 ucxx::Request::callback()  /src/ucxx/cpp/src/request.cpp:149
26 0x0000000000040014 ucp_request_complete_send()  /src/ucx-1.15/build/src/ucp/../../../src/ucp/core/ucp_request.inl:249
27 0x0000000000040014 ucp_ep_local_disconnect_progress()  /src/ucx-1.15/build/src/ucp/../../../src/ucp/core/ucp_ep.c:1625
28 0x000000000005e7f9 ucs_callbackq_slow_proxy()  /src/ucx-1.15/build/src/ucs/../../../src/ucs/datastruct/callbackq.c:404
29 0x000000000005fc4a ucs_callbackq_dispatch()  /src/ucx-1.15/build/../src/ucs/datastruct/callbackq.h:211
30 0x000000000005fc4a uct_worker_progress()  /src/ucx-1.15/build/../src/uct/api/uct.h:2777
31 0x000000000005fc4a ucp_worker_progress()  /src/ucx-1.15/build/src/ucp/../../../src/ucp/core/ucp_worker.c:2889
32 0x000000000005bebe ucxx::Worker::progressOnce()  /src/ucxx/cpp/src/worker.cpp:254
33 0x000000000005beec ucxx::Worker::progressPending()  /src/ucxx/cpp/src/worker.cpp:260
34 0x000000000005ca07 ucxx::Worker::progress()  /src/ucxx/cpp/src/worker.cpp:268
35 0x000000000005ca07 std::mutex::lock()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_mutex.h:100
36 0x000000000005ca07 std::lock_guard<std::mutex>::lock_guard()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_mutex.h:229
37 0x000000000005ca07 ucxx::Worker::progress()  /src/ucxx/cpp/src/worker.cpp:272
38 0x000000000004e1c1 std::_Function_handler<bool (), (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()::{lambda()#2}>::_M_invoke()  /src/ucxx/cpp/tests/listener.cpp:318
39 0x000000000004e1c1 std::__shared_ptr_access<ucxx::Request, (__gnu_cxx::_Lock_policy)2, false, false>::_M_get()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr_base.h:993
40 0x000000000004e1c1 std::__shared_ptr_access<ucxx::Request, (__gnu_cxx::_Lock_policy)2, false, false>::operator->()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/shared_ptr_base.h:987
41 0x000000000004e1c1 operator()()  /src/ucxx/cpp/tests/listener.cpp:319
42 0x000000000004e1c1 __invoke_impl<bool, (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()::<lambda()>&>()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/invoke.h:61
43 0x000000000004e1c1 __invoke_r<bool, (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()::<lambda()>&>()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/invoke.h:114
44 0x000000000004e1c1 _M_invoke()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_function.h:290
45 0x000000000007820d std::function<bool ()>::operator()()  /miniconda3/envs/rn-240313/x86_64-conda-linux-gnu/include/c++/11.4.0/bits/std_function.h:590
46 0x000000000005456a (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()  /src/ucxx/cpp/tests/listener.cpp:321
47 0x000000000005456a (anonymous namespace)::ListenerTest_EndpointNonBlockingCloseWithCallbacks_Test::TestBody()  /src/ucxx/cpp/tests/listener.cpp:321
48 0x000000000004d40e testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  ???:0
49 0x000000000004d6a1 testing::Test::Run()  ???:0
50 0x000000000004da2f testing::TestInfo::Run()  ???:0
51 0x000000000004deff testing::TestSuite::Run()  ???:0
52 0x0000000000059423 testing::internal::UnitTestImpl::RunAllTests()  ???:0
53 0x000000000004dfdd testing::UnitTest::Run()  ???:0
54 0x0000000000001072 main()  ???:0
55 0x0000000000024083 __libc_start_main()  /build/glibc-wuryBv/glibc-2.31/csu/../csu/libc-start.c:308
56 0x000000000002971e _start()  ???:0
=================================
Aborted
```

Reply:
Ah, ok, I didn't build in debug